Continuous Deployment

We want to move from big risky stressful releases to smaller less risky less stressfull releases.

We should use feature flags and other techniques that make new releases less risky.

In a spirit of continuous deployment, every commit should be published somewhere.

In the past we used to have "dev release" published on bintray, but bintray is dead and those release required a manual deployment/were not available for every commit.

https://jitpack.io/#jmfayard/refreshVersions could be a solution to do continuous deployment, but it currently doesn't work, details in the comments.

Other ideas:

- publish plugins like `de.fayard.refreshVersions.dev` with versions like `2021-05-10-1323.$gitSha`
- publish a `1.0-SNAPSHOT` version on mavenCentral

I created an issue for gradle plugin portal https://github.com/gradle/plugin-portal-requests/issues/98

**EDIT:** We will also need to update the releasing script.

In http4k, I tried to configure jitpack like this:

```kotlin
pluginManagement {
    val version = "0.10.0" 
    val jitpack = "com.github.jmfayard:refreshVersions:6d6eef3"
    repositories {
        gradlePluginPortal()
        if (jitpack.isNotBlank()) maven("https://jitpack.io")
    }
    resolutionStrategy {
        fun module(module: String) = when {
            jitpack.isNotBlank() -> jitpack
            else -> "de.fayard.refreshVersions:$module:$version"
        }
        eachPlugin {
            when (requested.id.id) {
                "de.fayard.refreshVersions" -> useModule(module("refreshVersions"))
                "de.fayard.buildSrcLibs" -> useModule(module("buildSrcLibs"))
            }
        }
    }
}

plugins {
    id("de.fayard.buildSrcLibs") 
    id("de.fayard.refreshVersions") 
}
```

but it fails hard

```
$ ./gradlew help
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all artifacts for configuration 'classpath'.
> Could not find de.fayard.refreshVersions:buildSrcLibs:0.10.0-LOCAL-SNAPSHOT.
  Searched in the following locations:
    - https://plugins.gradle.org/m2/de/fayard/refreshVersions/buildSrcLibs/0.10.0-LOCAL-SNAPSHOT/maven-metadata.xml
    - https://plugins.gradle.org/m2/de/fayard/refreshVersions/buildSrcLibs/0.10.0-LOCAL-SNAPSHOT/buildSrcLibs-0.10.0-LOCAL-SNAPSHOT.pom
    - https://jitpack.io/de/fayard/refreshVersions/buildSrcLibs/0.10.0-LOCAL-SNAPSHOT/maven-metadata.xml
    - https://jitpack.io/de/fayard/refreshVersions/buildSrcLibs/0.10.0-LOCAL-SNAPSHOT/buildSrcLibs-0.10.0-LOCAL-SNAPSHOT.pom
  Required by:
      unspecified:unspecified:unspecified > com.github.jmfayard:refreshVersions:6d6eef3
```

--- a/.github/workflows/publish-to-sonatype-snapshots.yml
+++ b/.github/workflows/publish-to-sonatype-snapshots.yml
@@ -1,0 +1,17 @@
+name: Publish to Sonatype Snapshots
+on:
+  workflow_dispatch:
+jobs:
+  build-and-upload:
+    name: Build and upload
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.4
+      - name: Build and upload artifacts
+        run: ./gradlew publishMavenPublicationToSonatypeSnapshotsRepository
+        env:
+          sonatype_username: ${{ secrets.SONATYPE_USERNAME }}
+          sonatype_password: ${{ secrets.SONATYPE_PASSWORD }}
+          GPG_key_id: ${{ secrets.GPG_KEY_ID }}
+          GPG_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          GPG_private_password: ${{ secrets.GPG_PRIVATE_PASSWORD }}

--- a/.github/workflows/release-plugins.yml
+++ b/.github/workflows/release-plugins.yml
@@ -20,3 +20,9 @@ jobs:
           wrapper-directory: plugins
           build-root-directory: plugins
           arguments: publishToAppropriateRepo --scan -Pgradle.publish.key=${{ secrets.gradle_publish_key }} -Pgradle.publish.secret=${{ secrets.gradle_publish_secret }}
+        env:
+          sonatype_username: ${{ secrets.SONATYPE_USERNAME }}
+          sonatype_password: ${{ secrets.SONATYPE_PASSWORD }}
+          GPG_key_id: ${{ secrets.GPG_KEY_ID }}
+          GPG_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          GPG_private_password: ${{ secrets.GPG_PRIVATE_PASSWORD }}

--- a/plugins/buildSrc/src/main/kotlin/Property.kt
+++ b/plugins/buildSrc/src/main/kotlin/Property.kt
@@ -1,0 +1,5 @@
+import org.gradle.api.provider.Property
+
+infix fun <T> Property<T>.by(value: T) {
+    set(value)
+}

--- a/plugins/buildSrc/src/main/kotlin/PropertyOrEnv.kt
+++ b/plugins/buildSrc/src/main/kotlin/PropertyOrEnv.kt
@@ -1,0 +1,11 @@
+import org.gradle.api.Project
+
+fun Project.propertyOrEnv(key: String): String {
+    return findProperty(key) as String?
+        ?: System.getenv(key)
+        ?: error("Didn't find any value for the key \"$key\" in Project properties or environment variables.")
+}
+
+fun Project.propertyOrEnvOrNull(key: String): String? {
+    return findProperty(key) as String? ?: System.getenv(key)
+}

--- a/plugins/buildSrc/src/main/kotlin/extensions/Project.kt
+++ b/plugins/buildSrc/src/main/kotlin/extensions/Project.kt
@@ -1,7 +1,0 @@
-package extensions
-
-import org.gradle.api.Project
-
-fun Project.propertyOrEnv(key: String): String? {
-    return findProperty(key) as String? ?: System.getenv(key)
-}

--- a/plugins/buildSrc/src/main/kotlin/publishing/Publishing.kt
+++ b/plugins/buildSrc/src/main/kotlin/publishing/Publishing.kt
@@ -1,8 +1,14 @@
 @file:Suppress("PackageDirectoryMismatch")
 
 import org.gradle.api.Project
+import org.gradle.api.UnknownTaskException
 import org.gradle.api.publish.PublishingExtension
 import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.jvm.tasks.Jar
+import org.gradle.kotlin.dsl.named
+import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
 
 object Publishing {
@@ -13,7 +19,13 @@ object Publishing {
 
 fun PublishingExtension.setupAllPublications(project: Project) {
     val mavenPublications = publications.withType<MavenPublication>()
-    mavenPublications.configureEach { setupPom() }
+    mavenPublications.configureEach {
+        artifact(project.tasks.emptyJavadocJar())
+        setupPom()
+    }
+    if (project.isSnapshot) {
+        sonatypeSnapshotsPublishing(project = project)
+    }
     project.registerPublishingTask()
     project.tasks.named("publishPlugins").configure {
         doFirst {
@@ -27,11 +39,21 @@ fun PublishingExtension.setupAllPublications(project: Project) {
     }
 }
 
+fun TaskContainer.emptyJavadocJar(): TaskProvider<Jar> {
+    val taskName = "javadocJar"
+    return try {
+        named(name = taskName)
+    } catch (e: UnknownTaskException) {
+        register(name = taskName) { archiveClassifier by "javadoc" }
+    }
+}
+
+
 private fun Project.registerPublishingTask() {
     require(project != rootProject)
 
     val publishTaskName = when {
-        isSnapshot -> "publishToMavenLocal"
+        isSnapshot -> "publishAllPublicationsToSonatypeSnapshotsRepository"
         else -> null // TODO: Replace with a dev repo making up for bintray?
     }
     val publishTask = publishTaskName?.let { tasks.named(it) }

--- a/plugins/buildSrc/src/main/kotlin/publishing/SonatypePublishing.kt
+++ b/plugins/buildSrc/src/main/kotlin/publishing/SonatypePublishing.kt
@@ -1,0 +1,44 @@
+@file:Suppress("PackageDirectoryMismatch")
+
+import org.gradle.api.Project
+import org.gradle.api.publish.PublishingExtension
+import java.net.URI
+
+
+fun PublishingExtension.mavenCentralStagingPublishing(
+    project: Project,
+    sonatypeUsername: String? = project.propertyOrEnvOrNull("sonatype_username"),
+    sonatypePassword: String? = project.propertyOrEnvOrNull("sonatype_password"),
+    repositoryId: String?
+) {
+    repositories {
+        maven {
+            name = "MavenCentralStaging"
+            url = when (repositoryId) {
+                null -> URI("https://oss.sonatype.org/service/local/staging/deploy/maven2/")
+                else -> URI("https://oss.sonatype.org/service/local/staging/deployByRepositoryId/$repositoryId/")
+            }
+            credentials {
+                username = sonatypeUsername
+                password = sonatypePassword
+            }
+        }
+    }
+}
+
+fun PublishingExtension.sonatypeSnapshotsPublishing(
+    project: Project,
+    sonatypeUsername: String? = project.propertyOrEnvOrNull("sonatype_username"),
+    sonatypePassword: String? = project.propertyOrEnvOrNull("sonatype_password")
+) {
+    repositories {
+        maven {
+            name = "SonatypeSnapshots"
+            url = URI("https://oss.sonatype.org/content/repositories/snapshots/")
+            credentials {
+                username = sonatypeUsername
+                password = sonatypePassword
+            }
+        }
+    }
+}

--- a/plugins/buildSrcLibs/build.gradle.kts
+++ b/plugins/buildSrcLibs/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
     id("com.gradle.plugin-publish")
     `java-gradle-plugin`
     `maven-publish`
+    signing
     `kotlin-dsl`
 }
 
@@ -23,6 +24,15 @@ pluginBundle {
     website = "https://jmfayard.github.io/refreshVersions"
     vcsUrl = "https://github.com/jmfayard/refreshVersions"
     tags = listOf("dependencies", "versions", "buildSrc", "kotlin", "kotlin-dsl")
+}
+
+signing {
+    useInMemoryPgpKeys(
+        propertyOrEnvOrNull("GPG_key_id"),
+        propertyOrEnvOrNull("GPG_private_key") ?: return@signing,
+        propertyOrEnv("GPG_private_password")
+    )
+    sign(publishing.publications)
 }
 
 publishing {

--- a/plugins/core/build.gradle.kts
+++ b/plugins/core/build.gradle.kts
@@ -5,6 +5,7 @@ plugins {
     `java-gradle-plugin`
     `java-test-fixtures`
     `maven-publish`
+    signing
     `kotlin-dsl`
     idea
 }
@@ -24,6 +25,15 @@ pluginBundle {
     website = "https://jmfayard.github.io/refreshVersions"
     vcsUrl = "https://github.com/jmfayard/refreshVersions"
     tags = listOf("dependencies", "versions", "buildSrc", "kotlin", "kotlin-dsl")
+}
+
+signing {
+    useInMemoryPgpKeys(
+        propertyOrEnvOrNull("GPG_key_id"),
+        propertyOrEnvOrNull("GPG_private_key") ?: return@signing,
+        propertyOrEnv("GPG_private_password")
+    )
+    sign(publishing.publications)
 }
 
 publishing {

--- a/plugins/dependencies/build.gradle.kts
+++ b/plugins/dependencies/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
     id("com.gradle.plugin-publish")
     `java-gradle-plugin`
     `maven-publish`
+    signing
     `kotlin-dsl`
 }
 
@@ -22,6 +23,15 @@ pluginBundle {
     website = "https://jmfayard.github.io/refreshVersions"
     vcsUrl = "https://github.com/jmfayard/refreshVersions"
     tags = listOf("dependencies", "versions", "buildSrc", "kotlin", "kotlin-dsl")
+}
+
+signing {
+    useInMemoryPgpKeys(
+        propertyOrEnvOrNull("GPG_key_id"),
+        propertyOrEnvOrNull("GPG_private_key") ?: return@signing,
+        propertyOrEnv("GPG_private_password")
+    )
+    sign(publishing.publications)
 }
 
 publishing {

